include/deal.II/grid/connectivity.h

```diff
@@ -0,0 +1,1477 @@
+      CRS<T> neighbors;
+
+      std::vector<unsigned int> quad_types;
+    };
+
+
+
+    /**
+     * Create the transposed of a compressed row storage matrix.
+     */
+    template <typename T>
+    CRS<T>
+    transpose(const CRS<T> &con)
+    {
+      auto        con_copy = con;
+      auto &      col      = con_copy.col;
+      const auto &ptr      = con_copy.ptr;
+
+      CRS<T> con_t;
+      auto & col_t = con_t.col;
+      auto & ptr_t = con_t.ptr;
+
+      for (unsigned int i = 0; i < ptr.size() - 1; i++)
+        std::sort(col.data() + ptr[i], col.data() + ptr[i + 1]);
+
+      col_t.clear();
+      ptr_t.clear();
+
+      col_t.reserve(col.size());
+      ptr_t.reserve(
+        col.size() == 0 ? 0 : *std::max_element(col.begin(), col.end()));
+
+      std::vector<std::pair<T, T>> temp;
+      temp.reserve(col.size());
+
+      for (unsigned int i = 0; i < ptr.size() - 1; i++)
+        for (std::size_t j = ptr[i]; j < ptr[i + 1]; j++)
+          temp.emplace_back(col[j], i);
+
+      std::sort(temp.begin(), temp.end());
+
+      T r = -1;
+
+      for (unsigned int i = 0; i < temp.size(); i++)
+        {
+          if (r != temp[i].first)
+            {
+              r = temp[i].first;
+              ptr_t.push_back(col_t.size());
+            }
+
+          col_t.push_back(temp[i].second);
+        }
+
+      ptr_t.push_back(col_t.size());
+
+      return con_t;
+    }
+
+
+
+    /**
+     * Determine the neighbors of a cell by visiting its faces and looking for
+     * a cell which is not equal to this cell.
+     */
+    template <typename T>
+    CRS<T>
+    determine_neighbors(const CRS<T> &con_cf, const CRS<T> &con_fc)
+    {
+      const auto &col_cf = con_cf.col;
+      const auto &ptr_cf = con_cf.ptr;
+      const auto &col_fc = con_fc.col;
+      const auto &ptr_fc = con_fc.ptr;
+
+      CRS<T> con_cc;
+      auto & col_cc = con_cc.col;
+      auto & ptr_cc = con_cc.ptr;
+
+      col_cc.clear();
+      ptr_cc.clear();
+
+      ptr_cc.push_back(0);
+
+      for (unsigned int i_0 = 0; i_0 < ptr_cf.size() - 1; i_0++)
+        {
+          for (std::size_t j_0 = ptr_cf[i_0]; j_0 < ptr_cf[i_0 + 1]; j_0++)
+            {
+              T temp = -1;
+
+              for (std::size_t j_1 = ptr_fc[col_cf[j_0]];
+                   j_1 < ptr_fc[col_cf[j_0] + 1];
+                   j_1++)
+                if (i_0 != col_fc[j_1])
+                  temp = col_fc[j_1];
+
+              col_cc.emplace_back(temp);
+            }
+
+          ptr_cc.push_back(col_cc.size());
+        }
+
+      return con_cc;
+    }
+
+
+
+    /**
+     * Determine the orientation of an entity of @p type described by its
+     * vertices @p var_1 relative to an entity described by @p var_0.
+     */
+    template <typename T, std::size_t N>
+    inline unsigned char
+    compute_orientation(const std::array<T, N> &var_0,
+                        const std::array<T, N> &var_1,
+                        const unsigned int      type)
```

Comment:
Can you put the `type` argument first (as it seems to control the rest of the arguments in terms of algorithm choice) and can you use more expressive variable names?